File: services/api/src/main/java/net/blueshell/api/mapping/MembershipMapper.java
package net.blueshell.api.mapping;


import net.blueshell.api.base.BaseMapper;
import net.blueshell.api.dto.FileDTO;
import net.blueshell.api.dto.MembershipDTO;
import net.blueshell.api.common.enums.FileType;
import net.blueshell.api.mapping.user.SimpleUserMapper;
import net.blueshell.api.model.File;
import net.blueshell.api.model.Membership;
import net.blueshell.api.service.UserService;
import org.mapstruct.*;
import org.springframework.beans.factory.annotation.Autowired;

import java.time.LocalDate;

@Mapper(componentModel = "spring", nullValuePropertyMappingStrategy = NullValuePropertyMappingStrategy.IGNORE)
public abstract class MembershipMapper extends BaseMapper<Membership, MembershipDTO> {

    @Autowired
    private FileMapper fileMapper;

    @Autowired
    private UserService userService;

    public abstract MembershipDTO toDTO(Membership membership);

    public abstract Membership fromDTO(MembershipDTO dto);

    @AfterMapping
    protected void afterFromDTO(MembershipDTO dto, @MappingTarget Membership membership) {
        if (dto.getSignature() != null) {
            FileDTO signatureDTO = dto.getSignature();
            signatureDTO.setFileType(FileType.SIGNATURE);
            File signature = fileMapper.fromDTO(dto.getSignature());
            membership.setSignature(signature);
        }
        if (membership.getStartDate() == null) {
            membership.setStartDate(LocalDate.now());
        }
        if (dto.getUserId() != null) {
            membership.setUser(userService.findById(dto.getUserId()));
        }
    }
}

File: api/src/main/kotlin/net/blueshell/api/membership/api/mapper/MembershipMapper.kt
package net.blueshell.api.membership.api.mapper

import net.blueshell.api.shared.enums.Role
import net.blueshell.api.membership.api.dto.MembershipDTO
import net.blueshell.api.shared.mapper.BaseMapper
import net.blueshell.api.membership.domain.model.Membership
import org.mapstruct.*


@Mapper(componentModel = "spring")
abstract class MembershipMapper : BaseMapper<Membership, MembershipDTO>() {
    @BeanMapping(ignoreByDefault = true)
    @Mapping(target = "id")
    @Mapping(target = "userId")
    @Mapping(target = "memberType")
    @Mapping(target = "startDate")
    @Mapping(target = "endDate")
    @Mapping(target = "incasso")
    @Mapping(target = "version")
    abstract override fun toDTO(membership: Membership): MembershipDTO

    @Mapping(target = "userId")
    @Mapping(target = "version")
    @BeanMapping(ignoreByDefault = true)
    abstract fun fromDTO(dto: MembershipDTO, @MappingTarget membership: Membership): Membership

    @AfterMapping
    protected fun afterFromDTO(dto: MembershipDTO, @MappingTarget membership: Membership) {
        if (!hasAuthority(Role.BOARD)) return

        dto.startDate?.let { membership.startDate = it }
        membership.endDate = dto.endDate // Must be applied, in order to be able to resume memberships
        dto.memberType?.let { membership.memberType = it }
        dto.incasso.let { membership.incasso = it }
    }
}
